lib/tools/system-calls.js

```diff
@@ -883,4 +883,75 @@ systemCallMethods.fileSize = async function (remotePath) {
   }
 };
 
+/**
+ * Installs the given certificate on a rooted real device or
+ * an emulator. The emulator must be executed with `-writable-system`
+ * command line option and adb daemon should be running in root
+ * mode for this method to work properly. The method also requires
+ * openssl tool to be available on the destination system.
+ * Read https://github.com/appium/appium/issues/10964
+ * for more details on this topic
+ *
+ * @param {Buffer|string} cert - base64-decoded content of the actual certificate
+ * represented as a string or a buffer
+ * @throws {Error} If openssl tool is not available on the destination system
+ * or if there was an error while installing the certificate
+ */
+systemCallMethods.installMitmCertificate = async function (cert) {
+  const binaryName = system.isWindows() ? 'openssl.exe' : 'openssl';
+  try {
+    await fs.which(binaryName);
+  } catch (err) {
+    throw new Error('To install a custom cert, the openssl tool must be installed ' +
+      'on the system and available on the path');
+  }
+
+  if (!_.isBuffer(cert)) {
+    cert = Buffer.from(cert, 'base64');
+  }
+
+  const tempCert = await tempDir.open({
+    prefix: 'cert',
+    suffix: '.cer'
+  });
+  try {
+    await fs.writeFile(tempCert.path, cert);
+    let {stdout} = await exec(binaryName, ['x509', '-noout', '-hash', '-in', tempCert.path]);
+    const certHash = stdout.trim();
+    log.debug(`Got certificate hash: ${certHash}`);
+    log.debug('Preparing certificate content');
+    ({stdout} = await exec(binaryName, ['x509', '-in', tempCert.path], {isBuffer: true}));
+    let dstCertContent = stdout;
+    ({stdout} = await exec(binaryName, ['x509',
+      '-in', tempCert.path,
+      '-text',
+      '-fingerprint',
+      '-noout'], {isBuffer: true}));
+    dstCertContent = dstCertContent.concat(stdout);
+    const dstCert = await tempDir.open({
+      prefix: certHash,
+      suffix: '.0'
+    });
+    try {
+      await fs.writeFile(dstCert.path, dstCertContent);
+      log.debug('Mounting /system endpoint');
+      await this.shell(['mount', '-o', 'rw,remount', '/system']);
+      const dstPath = '/system/etc/security/cacerts';
+      log.debug(`Uploading the generated certificate from '${dstCert.path}' to '${dstPath}'`);
+      await this.push(dstCert.path, dstPath);
+      log.debug('Remounting the remote file system');
+      await this.adbExec(['remount']);
+    } finally {
+      await fs.rimraf(dstCert.path);
```

Comment:
Does deleting this cert destroy the file descriptor? How is that resource handled?

Reply:
I don't get the question. The idea is to free the place as soon as possible, otherwise these temp files are going to be deleted, probably, after the next OS reboot.

Reply:
The question is in regards to `dstCert`, which is created by [tempDir.open](https://github.com/appium/appium-support/blob/master/lib/tempdir.js#L32-L42). This means that the file is open. I'm wondering if, in practice, running `fs.rimraf` frees all those resources.

Reply:
@imurchie I'm unable to find the exact description of what is going to happen in such case (perhaps, this depends on low-level OS implementation), but I assume it would be safer to call `fs.close` before doing file removal.